### PR TITLE
Fix allocated node counter and open/close node linked lists

### PR DESCRIPTION
**Pull request by sergiosota**
Hi, I'm Sergio Sota, from Spain. This is my very first pull request. Sorry for my terrible English :-P I have used your astar algorithm implementation in a bot for an AI contest (check here https://github.com/marcan/lighthouses_aicontest) that will take place this july on Bilbao (Euskal Encounter 26)

The thing is... sometimes somewhere some bot hung (I run many instances of it in parallel with different parameters in order to train) So I begun the bug hunt... it took me almost 2 weeks but finally I think I have found the problem.

1) When there is no more memory (I used the default 1000 node size) the program tries to free all nodes: the open list, the closed list, the goal node... but there is another node that doesn't get freed: the "n" node (the one pulled out from the open list at the beginning and used here and there) I used EnsureMemoryFree method assertion just after that and the program stops (it's not big deal since the astar object gets destroyed once you finish your search)

2) If you select not to use FSA memory allocator... when a node gets freed there is a "m_AllocateNodeCount--" but when a node gets allocated there is no "m_AllocateNodeCount++"

3) This is, I believe, the most important issue: all nodes, once allocated in memory, have references to other nodes... so we can't realloc any of them or the solution chain could be broken. e.g. (A) -> (B) -> (C) if we realloc B to other place it will still point to C, but now A is pointing to the position B occupied. If FSA doesn't use that block then we are lucky, but if it is somehow refilled with other block D... then we have (A) -> (D) -> (?)
In other words... when we have a successor that will replace other old open/close node (because this successor has a better path) we must change the old/close node in place (because other node could be pointing to it) Right now the successor is pulled in as a new node and the old open/closed node is erased... and that could broke the chain.

Well, I have tested the modified version and it is working ok so far! Thanks for you time and for your great project :-)

### CODE REVIEW

**Comment by justinhj** [comment]
Hi Sergio. Many thanks for using my code and for spending all that time diagnosing and fixing this issue! I will spend some time reviewing the code this weekend.

**Comment by justinhj** [comment]
I have checked the changes and didn't see any problems.
Thanks again, Sergio

**Comment by sergiosota** [comment]
Great! Thank you! :-D
